exporters/prometheus/exporter.go

```diff
@@ -0,0 +1,185 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package prometheus // import "github.com/open-telemetry/opentelemetry-go/exporters/prometheus"
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+)
+
+// Exporter is a Prometheus exporter.
+type Exporter struct {
+	reader   metric.Reader
+	registry *prometheus.Registry
+}
+
+// NewPrometheusExporter returns a Prometheus exporter.
+func NewPrometheusExporter(reader metric.Reader) (*Exporter, error) {
+	e := &Exporter{
+		reader:   reader,
+		registry: prometheus.NewRegistry(),
+	}
+
+	if err := e.registry.Register(e); err != nil {
+		return nil, err
+	}
+
+	return e, nil
+}
+
+// Describe implements prometheus.Collector.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	metrics, err := e.reader.Collect(context.TODO())
+	if err != nil {
+		otel.Handle(err)
+	}
+	for _, metricData := range getMetricData(metrics) {
+		ch <- metricData.description
+	}
+}
+
+// Collect implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	metrics, err := e.reader.Collect(context.TODO())
+	if err != nil {
+		otel.Handle(err)
+	}
+
+	for _, metricData := range getMetricData(metrics) {
+		if metricData.valueType == prometheus.UntypedValue {
+			m, err := prometheus.NewConstHistogram(metricData.description, metricData.histogramCount, metricData.histogramSum, metricData.histogramBuckets, metricData.attributeValues...)
+			if err != nil {
+				otel.Handle(err)
+			}
+			ch <- m
+		} else {
+			m, err := prometheus.NewConstMetric(metricData.description, metricData.valueType, metricData.value, metricData.attributeValues...)
+			if err != nil {
+				otel.Handle(err)
+			}
+			ch <- m
+		}
+	}
+}
+
+// metricData holds the metadata as well as values for individual data points.
+type metricData struct {
+	name             string
+	description      *prometheus.Desc
+	attributeValues  []string
+	valueType        prometheus.ValueType
+	value            float64
+	histogramCount   uint64
+	histogramSum     float64
+	histogramBuckets map[float64]uint64
+}
+
+func getMetricData(metrics metricdata.ResourceMetrics) []*metricData {
+	allMetrics := make([]*metricData, 0)
+	for _, scopeMetrics := range metrics.ScopeMetrics {
+		for _, m := range scopeMetrics.Metrics {
+			switch v := m.Data.(type) {
+			case metricdata.Histogram:
+				allMetrics = append(allMetrics, getHistogramMetricData(v, m)...)
+			case metricdata.Sum[int64]:
```

Comment:
curious: is it possible to combine the two sum cases into a `case metricdata.Sum`?

Reply:
that's what I actually had originally, using generics like:
```go
func getMetricData[N int64 | float64](..) {
...
  case metricdata.Sum[N]:
}

but go couldn't infer the type for N without explicitly passing the type to `getMetricData`. maybe there's a better way, but I'm new to generics

Reply:
I think its fine to leave as-is here.